sql: use the new opt index constraints code

RaduBerinde opened the pull request:
This is a strawman proposal on how to proceed. I left the existing code in place and added the new code as "experimental", controlled by a flag. This will allow reverting to the old code if necessary; once we don't think this will be needed, we can nuke old code.

#### opt: support placeholders

Evaluate placeholders when building a scalar expression.

Release note: None

#### opt: support sub-operators (for SOME/ALL)

This change is required to properly pass through expressions like
`x < SOME(SELECT y FROM ...)`. In this case, the operator is someOp
and the suboperator is ltOp.

Includes a necessary fix in ComparisonExpr to tolerate TTuple
arguments (the type of the subquery clause).

Release note: None

#### sql: use the new opt index constraints code

The change installs an "experimental" path which uses the new index
constraints code in `opt`. It is controlled by a flag so we can turn
it off if necessary without a big code change. The intention is to
clean up the old code once we are convinced the new code is stable. I

As far as I can tell, this addresses all the problems described
in #6346 and associated issues. I plan to add logictests for
these cases before closing the issues.

The logic tests changes are due to one of the following:

 - unnecessary NULL constraints: the old code did a moderate effort to
   keep spans simple and don't use `/!NULL` constraints when not
   necessary. The new code does a better job; it never generates such
   constraints on non-nullable columns. So a lot of tests show changes
   like `/!NULL-/4` -> ~-/4` but these are equivalent when the column
   is not nullable.

 - improvements: in some cases the new code generates better (more
   constrained) spans. One improvement here is that the new code also
   generates constraints on implicit columns (this is part of #6346, I
   will add a test showing this). Another improvement is that in some
   cases we do a better job at simplifying the remaining filter (in
   particular for tuple inequalities).

 - there is one case where we used to detect a contradiction but we
   now don't.  The reason is that the contradiction is on a column
   that is not part of any index, and the new code only looks at index
   columns. In the future, we may want to precalculate single-column
   constraints and reuse them acrosss all indexes, and in that case we
   will detect it again.

 - there is a bug in the old index selection code where we have a
   disjunction with one of the expressions having a subquery and not
   generating constraints. This should result in a full table scan
   but the old code incorrectly generates a span. I looked at the old
   code but I can't find an easy way to fix it (or even detect if we
   might be hitting it). A big part of why we have rewritten all
   this code is because the old code is hard to understand and fix or
   improve.

Release note (performance improvement): more efficient scans for certain
WHERE clauses.

Planning benchmarks:
```
name                                                             old time/op    new time/op    delta
Planning/SELECT_*_FROM_abc-8                                        228µs ± 5%     228µs ± 3%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                 155µs ± 0%     157µs ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                            151µs ± 0%     152µs ± 2%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                 161µs ± 0%     161µs ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                            154µs ± 2%     152µs ± 1%   ~     (p=0.667 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8     385µs ± 4%     369µs ± 1%   ~     (p=0.667 n=2+2)

name                                                             old alloc/op   new alloc/op   delta
Planning/SELECT_*_FROM_abc-8                                       29.5kB ± 0%    29.5kB ± 0%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                20.6kB ± 1%    21.9kB ± 1%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                           21.7kB ± 0%    23.1kB ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                23.0kB ± 1%    24.3kB ± 1%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                           21.7kB ± 0%    23.1kB ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8    65.2kB ± 0%    65.2kB ± 0%   ~     (p=1.000 n=2+2)

name                                                             old allocs/op  new allocs/op  delta
Planning/SELECT_*_FROM_abc-8                                          308 ± 0%       308 ± 0%   ~     (all equal)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                   320 ± 0%       316 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                              306 ± 0%       304 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                   350 ± 0%       348 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                              304 ± 0%       302 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8       601 ± 0%       601 ± 0%   ~     (all equal)
```

RaduBerinde committed:
opt: support placeholders

Evaluate placeholders when building a scalar expression.

Release note: None

RaduBerinde committed:
opt: support sub-operators (for SOME/ALL)

This change is required to properly pass through expressions like
`x < SOME(SELECT y FROM ...)`. In this case, the operator is someOp
and the suboperator is ltOp.

Includes a necessary fix in ComparisonExpr to tolerate TTuple
arguments (the type of the subquery clause).

Release note: None

RaduBerinde committed:
sql: use the new opt index constraints code

The change installs an "experimental" path which uses the new index
constraints code in `opt`. It is controlled by a flag so we can turn
it off if necessary without a big code change. The intention is to
clean up the old code once we are convinced the new code is stable. I

As far as I can tell, this addresses all the problems described
in #6346 and associated issues. I plan to add logictests for
these cases before closing the issues.

The logic tests changes are due to one of the following:

 - unnecessary NULL constraints: the old code did a moderate effort to
   keep spans simple and don't use `/!NULL` constraints when not
   necessary. The new code does a better job: it never generates such
   constraints on non-nullable columns. So a lot of tests show changes
   like `/!NULL-/4` -> ~-/4` but these are equivalent when the column
   is not nullable.

 - improvements: in some cases the new code generates better (more
   constrained) spans. One improvement here is that the new code also
   generates constraints on implicit columns (this is part of #6346, I
   will add a test showing this). Another improvement is that in some
   cases we do a better job at simplifying the remaining filter (in
   particular for tuple inequalities).

 - there is one case where we used to detect a contradiction but we
   now don't.  The reason is that the contradiction is on a column
   that is not part of any index, and the new code only looks at index
   columns. In the future, we may want to precalculate single-column
   constraints and reuse them acrosss all indexes, and in that case we
   will detect it again.

 - there is a bug in the old index selection code where we have a
   disjunction with one of the expressions having a subquery and not
   generating constraints. This should result in a full table scan
   but the old code incorrectly generates a span. I looked at the old
   code but I can't find an easy way to fix it (or even detect if we
   might be hitting it). A big part of why we have rewritten all
   this code is because the old code is hard to understand and fix or
   improve.

Release note (performance improvement): more efficient scans for certain
WHERE clauses.

Planning benchmarks:
```
name                                                             old time/op    new time/op    delta
Planning/SELECT_*_FROM_abc-8                                        228µs ± 5%     228µs ± 3%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                 155µs ± 0%     157µs ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                            151µs ± 0%     152µs ± 2%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                 161µs ± 0%     161µs ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                            154µs ± 2%     152µs ± 1%   ~     (p=0.667 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8     385µs ± 4%     369µs ± 1%   ~     (p=0.667 n=2+2)

name                                                             old alloc/op   new alloc/op   delta
Planning/SELECT_*_FROM_abc-8                                       29.5kB ± 0%    29.5kB ± 0%   ~     (p=1.000 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                20.6kB ± 1%    21.9kB ± 1%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                           21.7kB ± 0%    23.1kB ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                23.0kB ± 1%    24.3kB ± 1%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                           21.7kB ± 0%    23.1kB ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8    65.2kB ± 0%    65.2kB ± 0%   ~     (p=1.000 n=2+2)

name                                                             old allocs/op  new allocs/op  delta
Planning/SELECT_*_FROM_abc-8                                          308 ± 0%       308 ± 0%   ~     (all equal)
Planning/SELECT_*_FROM_abc_WHERE_a_>_5_ORDER_BY_a-8                   320 ± 0%       316 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5-8                              306 ± 0%       304 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_b_=_5_ORDER_BY_a-8                   350 ± 0%       348 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_WHERE_c_=_5-8                              304 ± 0%       302 ± 0%   ~     (p=0.333 n=2+2)
Planning/SELECT_*_FROM_abc_JOIN_abc_AS_abc2_ON_abc.a_=_abc2.a-8       601 ± 0%       601 ± 0%   ~     (all equal)
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21217)
<!-- Reviewable:end -->

knz commented:
Architecture proposal LGTM, I'm happy about how little code changes there are. Need more insight about why some tests are changing though.

---

Reviewed 3 of 3 files at r1, 3 of 3 files at r2, 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 140 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y5u6ynfNKjsnBR-qJ:-L1y5u6zIGQAx7lRGgOD:b-uwj8bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L140)):*
> ```Go
> 				}
> 				if spans, ok := c.ic.Spans(); ok && len(spans) == 0 {
> 					// No spans (i.e. the filter is always false).
> ```

Extend the comment to indicate that for filters where the span cannot be determined reliably, the constraint generator would generate a full span (/keybegin-/keyend). So lack of span indeed means 0 rows selected.

---

*[pkg/sql/opt_index_selection.go, line 1187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6CZJXzXe_mTjcDv7:-L1y6CZKCp_rRb_VXZ3f:bcgjvre) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L1187)):*
> ```Go
> ) (roachpb.Spans, error) {
> 	spans := make(roachpb.Spans, len(logicalSpans))
> 	interstices := make([][]byte, len(index.ColumnDirections)+len(index.ExtraColumnIDs)+1)
> ```

Why this change?

---

*[pkg/sql/logictest/testdata/logic_test/explain_types, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6kWW2MBXHebPs4_C:-L1y6kWW2MBXHebPs4_D:bri4wkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/explain_types#L127)):*
> ```
> ·                         4  ·       table        t@primary                                                                  ·               ·
> ·                         4  ·       spans        ALL                                                                        ·               ·
> ·                         4  ·       filter       ((((v)[int] > (123)[int])[bool]) AND (((v)[int] < (2)[int])[bool]))[bool]  ·               ·
> ```

huh why did this change? I think the span should be empty and the source still reduced to zeroNode.

---

*[pkg/sql/logictest/testdata/logic_test/join, line 1113 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6zrOMUQ7ZchLOO4q:-L1y6zrOMUQ7ZchLOO4r:bv6hzes) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/join#L1113)):*
> ```
>            └── scan  ·         ·
> ·                    table     square@primary
> ·                    spans     -/5/#
> ```

The predicate `n < 6` in the join predicate implies `n is not null`, so I think the span here is too large.

---

*[pkg/sql/logictest/testdata/logic_test/join, line 1668 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7AAwK6a0B-8P9B1U:-L1y7AAwK6a0B-8P9B1V:b-5a6877) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/join#L1668)):*
> ```
>  ├── scan  1  scan  ·               ·                  (x, y, u, x, y, v)  x=x; y=y; x=CONST; y!=NULL
>  │         1  ·     table           xyu@primary        ·                   ·
>  │         1  ·     spans           /1-/1/10           ·                   ·
> ```

ditto I think and the other similar spans below

---

*[pkg/sql/logictest/testdata/logic_test/join, line 1909 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7N2gmzP88sVbzgH1:-L1y7N2hCePTjpxkkV2R:b-6n7kbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/join#L1909)):*
> ```
>       ├── scan  2  scan    ·               ·                  (x, y, u, v)  x!=NULL; y!=NULL
>       │         2  ·       table           xyu@primary        ·             ·
>       │         2  ·       spans           /1/2/4-            ·             ·
> ```

this is correct though I think

---

*[pkg/sql/logictest/testdata/logic_test/select_index, line 163 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7QABl6l7LLpfxCe4:-L1y7QABl6l7LLpfxCe5:b-hvba9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_index#L163)):*
> ```
>                 ├── scan   ·               ·
>                 │          table           t@bc
>                 │          spans           -/"3"
> ```

here again I think it implies b is not null

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 192 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7bDBXdf0jEkHPg0B:-L1y7bDBXdf0jEkHPg0C:b-urtpwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L192)):*
> ```
> scan  ·      ·
> ·     table  p1@b
> ·     spans  -/4
> ```

ditto here and below

---

*[pkg/sql/logictest/testdata/logic_test/subquery, line 490 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y5gD4qaUC1PRk3RKd:-L1y5gD5t6_cUDh6yFZD:b-r0i2en) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/subquery#L490)):*
> ```
> ----
> render               ·           ·
>  └── scan            ·           ·
> ```

I'm confused, how can you have a render as child to a scan?

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 490 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7mm1fNQN_MatMcwt:-L1y7mm1fNQN_MatMcwu:biuimdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L490)):*
> ```
>  └── scan  1  scan    ·         ·                       (u, v, w)  +u,+v,+w
> ·          1  ·       table     uvw@uvw_u_v_w_idx       ·          ·
> ·          1  ·       spans     -/1/2/3 /1/2/4-         ·          ·
> ```

very nice!

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 491 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7oLnXe_0c5tLXzjX:-L1y7oLnXe_0c5tLXzjY:b-b8q8ui) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L491)):*
> ```
> ·          1  ·       table     uvw@uvw_u_v_w_idx       ·          ·
> ·          1  ·       spans     -/1/2/3 /1/2/4-         ·          ·
> ·          1  ·       filter    (u, v, w) != (1, 2, 3)  ·          ·
> ```

why didn't this filter disappear

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1y7rxDM3dGD0f6Urc1:b8vcpmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
> ```
> ----
> NULL  NULL  1
> NULL  NULL  2
> ```

welp, what's going on here

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 609 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y83aMJ8nZI6VwsyZj:-L1y83aMJ8nZI6VwsyZk:b-7fmgbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L609)):*
> ```
>  └── scan  1  scan    ·         ·                  (u, v, w)  +u,+v,+w
> ·          1  ·       table     uvw@uvw_u_v_w_idx  ·          ·
> ·          1  ·       spans     /!NULL-/2          ·          ·
> ```

very nice too

---

*[pkg/sql/logictest/testdata/logic_test/update, line 488 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y85YNj9uj6WqsoFEh:-L1y85YNj9uj6WqsoFEi:bjlueyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/update#L488)):*
> ```
>            └── scan  ·      ·
> ·                    table  kv@primary
> ·                    spans  -/2/#
> ```

again I think k is not null is implied here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L1y8B80aXO9ED9qAnW9:bls3f7l)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/subquery, line 490 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y5gD4qaUC1PRk3RKd:-L1y8fR7UXeiL40_9vUR:bu3pegc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/subquery#L490)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm confused, how can you have a render as child to a scan?
</blockquote></details>

ignore that comment, there's a subquery. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

knz commented:
In reaction to 

> Not sure if its worth fixing the old code if it's going away

Don't make a single PR that merges a bug fix with a general code change meant to keep tests as-is. If there are bug in the current code, mark the tests as incorrect (and, preferably, add a failing assertion in the planning code to reject queries where w know there's a problem, if you don't plan to fix them), and then in the commit where you add the new code, show that the tests are left unchanged.

---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L1y8sJg3MIsbO7XRU4x:b-t1fxuu)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I'm not merging a bug fix with another change. We're switching to a new subsystem which does not have the same bugs as the existing subsystem. I will fix the issue in the old code (though IMO it's overkill if it just for the cosmetics of this PR)

---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 1187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6CZJXzXe_mTjcDv7:-L1yN1S63UgX4_03d0oe:b-op445z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L1187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why this change?
</blockquote></details>

The old code did not generate constraints on implicit columns (even though it could/should have). See https://github.com/cockroachdb/cockroach/issues/6346#issuecomment-354118411

---

*[pkg/sql/logictest/testdata/logic_test/explain_types, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6kWW2MBXHebPs4_C:-L1yOW59wj5P9BnJ6aUl:b-hbhjhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/explain_types#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

huh why did this change? I think the span should be empty and the source still reduced to zeroNode.
</blockquote></details>

`v` is not a column in any index. We can only use constraints on index columns, other columns can't be represented in spans. The old code just happened to detect the contradiction.

---

*[pkg/sql/logictest/testdata/logic_test/join, line 1113 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6zrOMUQ7ZchLOO4q:-L1yOoM_yrYIH3NT1fks:bbb141s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/join#L1113)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The predicate `n < 6` in the join predicate implies `n is not null`, so I think the span here is too large.
</blockquote></details>

And the fact that it's a PK column also implies `n is not null` :) The old code did a moderate effort to keep spans simple and avoid unnecessary `!NULL`s; the new code does a better job.

---

*[pkg/sql/logictest/testdata/logic_test/select_index, line 163 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7QABl6l7LLpfxCe4:-L1yPTS4_khggGF966K7:b-4d7kx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_index#L163)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

here again I think it implies b is not null
</blockquote></details>

b is a PK column (not nullable). Same below

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 491 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7oLnXe_0c5tLXzjX:-L1yPaG8yCQlrpCMZVeD:b-gsktbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L491)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

why didn't this filter disappear
</blockquote></details>

Because we still need to exclude results like `1, NULL, 3` or `1, 2, NULL`.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1yPkbyyVPum0EiIdc5:b4kdxnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

welp, what's going on here
</blockquote></details>

This was broken before. All these results are valid.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L1yNuWmq2CGJURMKu7z:b-z31afd)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 1187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6CZJXzXe_mTjcDv7:-L1yRw4iRT17yk8LDLtb:b-82duz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L1187)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The old code did not generate constraints on implicit columns (even though it could/should have). See https://github.com/cockroachdb/cockroach/issues/6346#issuecomment-354118411
</blockquote></details>

Oh interesting. Is this new logic explained somewhere?

---

*[pkg/sql/logictest/testdata/logic_test/explain_types, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6kWW2MBXHebPs4_C:-L1yS4Y4NRjRwcqoZWaQ:bpgm2py) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/explain_types#L127)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`v` is not a column in any index. We can only use constraints on index columns, other columns can't be represented in spans. The old code just happened to detect the contradiction.
</blockquote></details>

This makes me think we could generate the spans for all columns not just index columns, and simplify the predicate accordingly if the computed span(s) is simpler than the original expression. Is there a fundamental reason why this would be a bad idea?

---

*[pkg/sql/logictest/testdata/logic_test/join, line 1113 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6zrOMUQ7ZchLOO4q:-L1ySK50PZBn4KyiPrdf:bdpswx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/join#L1113)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

And the fact that it's a PK column also implies `n is not null` :) The old code did a moderate effort to keep spans simple and avoid unnecessary `!NULL`s; the new code does a better job.
</blockquote></details>

All right, thanks for explaining.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 491 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7oLnXe_0c5tLXzjX:-L1ySTevmjdQzxU8kzLP:b-k73z76) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L491)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Because we still need to exclude results like `1, NULL, 3` or `1, 2, NULL`.
</blockquote></details>

thanks

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1ySVzYAVRofkh_e3MK:b-9wgb0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This was broken before. All these results are valid.
</blockquote></details>

Yeah so that's what I meant with bug fix != code subsitution

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 1187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6CZJXzXe_mTjcDv7:-L1yVo0yRf5vevgjM4sH:b1nhdt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L1187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Oh interesting. Is this new logic explained somewhere?
</blockquote></details>

No, there isn't anything different between "regular" and implicit columns in index selection. The fact that the old code didn't include them feels like an accidental omission.

---

*[pkg/sql/logictest/testdata/logic_test/explain_types, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6kWW2MBXHebPs4_C:-L1yVyd2VPxFTA6RYKqV:b-qts95s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/explain_types#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This makes me think we could generate the spans for all columns not just index columns, and simplify the predicate accordingly if the computed span(s) is simpler than the original expression. Is there a fundamental reason why this would be a bad idea?
</blockquote></details>

We can generate single-column spans and then reuse those for all the indexes. I might explore that as a future improvement. But we can't do it in general because of multicolumn spans (e.g. tuple inequalities). So it would be a hybrid approach where we pre-generate single-column spans, and then we generate multi-column spans for each index.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1yWar0t2JF9K8e82Kn:bxiz6d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yeah so that's what I meant with bug fix != code subsitution
</blockquote></details>

The bug fix comes as a result of the code substitution. I cannot substitute the code without fixing the bug. Like I said, I will fix the bug in the old code separately, but I maintain that as a general principle it's unreasonable to have to do that if the old code will just be deleted soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 1187 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6CZJXzXe_mTjcDv7:-L1yaUucJWIyzN7MX_50:b-dc62dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L1187)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

No, there isn't anything different between "regular" and implicit columns in index selection. The fact that the old code didn't include them feels like an accidental omission.
</blockquote></details>

:+1: thanks

---

*[pkg/sql/logictest/testdata/logic_test/explain_types, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y6kWW2MBXHebPs4_C:-L1yaaUvRIg8hYuHJ2cC:b-oz6uks) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/explain_types#L127)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can generate single-column spans and then reuse those for all the indexes. I might explore that as a future improvement. But we can't do it in general because of multicolumn spans (e.g. tuple inequalities). So it would be a hybrid approach where we pre-generate single-column spans, and then we generate multi-column spans for each index.
</blockquote></details>

yep I agree

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1yacGw9ycfB6nwdda7:bz1qtai) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The bug fix comes as a result of the code substitution. I cannot substitute the code without fixing the bug. Like I said, I will fix the bug in the old code separately, but I maintain that as a general principle it's unreasonable to have to do that if the old code will just be deleted soon.
</blockquote></details>

Look unless you know for certainty that soon == before three weeks from now (so we get enough testing time) we have to work with the expectation that the two code paths will live side by side in 2.0. 
Do you know this already?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/tuple, line 497 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y7rxDM3dGD0f6Urc0:-L1yba4DZAAedpW_l69o:b22rmsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/tuple#L497)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Look unless you know for certainty that soon == before three weeks from now (so we get enough testing time) we have to work with the expectation that the two code paths will live side by side in 2.0. 
Do you know this already?
</blockquote></details>

Fair point. I think there's a very high likelihood that 2.0 will use the new code, but of course I can't be 100%.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes agree with your assessment of likelihood but let's make the conservative engineering choice. 

Also if there's a patch to fix broken queries I'll want to cherry-pick that onto 1.1.

rytaft commented:
As a sanity check, I think it would be good to also test the path in which `useExperimentalIndexConstraints` is false.  In the worst case that you need to revert to the old code, can you be sure it will still work?

To echo @knz, can you add a bit more info to the commit message about why the tests have changed?  E.g., why is `/!NULL` no longer included in the output spans?

I've looked through half of the files - will come back and do the other half in a bit.

---

Reviewed 6 of 11 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt_index_selection.go, line 2026 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L20s6COBAarU5g2X8zJ:-L20s6COBAarU5g2X8zK:b-bpt1fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L2026)):*
> ```Go
> 	}
> 	if !ok {
> 		v.cost *= 1000
> ```

Can you add a comment here to explain why you're multiplying cost by 1000?  What does it mean if ok is false?

---

*[pkg/sql/opt_index_selection.go, line 2151 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L20yrUsfAVlepy3o80L:-L20yrUsfAVlepy3o80M:b-d86zs7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L2151)):*
> ```Go
> 	// We tighten the end key to prevent reading interleaved children
> 	// after the last parent key.
> 	s.EndKey, err = sqlbase.AdjustEndKeyForInterleave(tableDesc, index, s.EndKey, ls.End.Inclusive)
> ```

Does this function also adjust the span based on ls.End.Inclusive? Maybe include this info in the comment...

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L214NBvl8kXmBMe4_Hj:-L214NBvl8kXmBMe4_Hk:b-34ogt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L246)):*
> ```
> ----
> scan  ·      ·
> ·     table  p1@primary
> ```

Why did the choice of index change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L212BekUaZRHnt8MQ-8:b-1q8baq)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Good point, I ran the logictests with the flag set to false. Updated the commit message as well.

---

Review status: 2 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt_index_selection.go, line 140 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L1y5u6ynfNKjsnBR-qJ:-L21B8lIQbVCyNOR8SAR:b-w1kw8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L140)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Extend the comment to indicate that for filters where the span cannot be determined reliably, the constraint generator would generate a full span (/keybegin-/keyend). So lack of span indeed means 0 rows selected.
</blockquote></details>

If there are no constraints, `ok` is false. This is documented for `Spans()`. Included it in the comment here as well.

---

*[pkg/sql/opt_index_selection.go, line 2026 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L20s6COBAarU5g2X8zJ:-L21BRvsOkk50C32YbTp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L2026)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment here to explain why you're multiplying cost by 1000?  What does it mean if ok is false?
</blockquote></details>

Done.

---

*[pkg/sql/opt_index_selection.go, line 2151 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L20yrUsfAVlepy3o80L:-L21BT6jyUjLHcbWpSZv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/opt_index_selection.go#L2151)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Does this function also adjust the span based on ls.End.Inclusive? Maybe include this info in the comment...
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L214NBvl8kXmBMe4_Hj:-L217JXx6_GHX8lq4aU7:b-hcu67c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L246)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why did the choice of index change?
</blockquote></details>

Just an artifact of how we calculate the costs. We use the fraction of columns which have constraints, and the new code includes implicit columns (the non-unique index on `b` here is similar to an index `(b, a)`; the old code treated it just like a unique index on `b`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L216gTwWAaDIlODPNsP:b-63sokp)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I just skimmed this change, but :thumbsup: that this is ready to go.

---

Reviewed 3 of 3 files at r1, 3 of 3 files at r2.
Review status: 2 of 16 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 195 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L210byZTLzpb-bcdyW3:-L210byZTLzpb-bcdyW4:b-q64et5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37c602416401f4ebad8f73318e8a82c17cdb95/pkg/sql/opt/build.go#L195)):*
> ```Go
> 			bc.buildScalar(t.TypedRight()),
> 		)
> 		e.subOperator = comparisonOpMap[t.SubOperator]
> ```

I recall reading that `= ANY` is equivalent to `IN`. Do we do that normalization somewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L21ZMW0EPjgD4mDaJ1N:b-tlisc0)*
<!-- Sent from Reviewable.io -->

knz commented:
LGTM too

RaduBerinde commented:
*[pkg/sql/opt/build.go, line 195 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L210byZTLzpb-bcdyW3:-L21fUO0D5Cpb8jOTNvp:b-8pxirj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37c602416401f4ebad8f73318e8a82c17cdb95/pkg/sql/opt/build.go#L195)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I recall reading that `= ANY` is equivalent to `IN`. Do we do that normalization somewhere?
</blockquote></details>

I don't think so. There's no SubOperator usage in tree/normalize.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 16 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 195 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L210byZTLzpb-bcdyW3:-L21jXi-B-ECc4INOO7F:b-vge8d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37c602416401f4ebad8f73318e8a82c17cdb95/pkg/sql/opt/build.go#L195)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think so. There's no SubOperator usage in tree/normalize.go
</blockquote></details>

File an issue or add a TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 195 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L210byZTLzpb-bcdyW3:-L21oxyjt4Po41zQQFHv:bbmrvkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37c602416401f4ebad8f73318e8a82c17cdb95/pkg/sql/opt/build.go#L195)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

File an issue or add a TODO.
</blockquote></details>

#21240

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 195 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L210byZTLzpb-bcdyW3:-L21pTES8EZHa2MIaApB:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f37c602416401f4ebad8f73318e8a82c17cdb95/pkg/sql/opt/build.go#L195)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

#21240
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  

[nit] in the commit message you say "the old code does a better job".  I think you meant new code?

---

Reviewed 2 of 3 files at r1, 3 of 3 files at r5, 2 of 11 files at r6.
Review status: 7 of 16 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/select_tighten_spans, line 246 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L214NBvl8kXmBMe4_Hj:-L21o_N1fc8VjD0x4i1_:b513kxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f69c8a6368e2fc897f0dd124bdbefb51fdd14c2/pkg/sql/logictest/testdata/logic_test/select_tighten_spans#L246)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Just an artifact of how we calculate the costs. We use the fraction of columns which have constraints, and the new code includes implicit columns (the non-unique index on `b` here is similar to an index `(b, a)`; the old code treated it just like a unique index on `b`).
</blockquote></details>

Thanks for the info!

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L21sWeEFbrZPskYikjV:b-rgaabe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
> ```Go
> 			rightRet = t.Typ
> 		case types.TTuple:
> 			rightRet = t[0]
> ```

Can you add a comment to explain this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L21nogN7XVdHV9JkjqT:bfb1rs5)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 7 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L21uQwDqvQ4wOo3Ro7f:b-fdh09p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment to explain this change?
</blockquote></details>

@petermattis @knz  can you help me explain this? I don't quite know why `TArray` was used here at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 7 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L21vjQQxfSRGkrTcsPA:b-xd686g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

@petermattis @knz  can you help me explain this? I don't quite know why `TArray` was used here at all.
</blockquote></details>

Subqueries are typed as `TTuple`. The type checking code allows `{ANY,SOME,ALL} (<tuple>)` in order to support `SELECT 1 = ANY(SELECT 1, 2)`. 

Note that cockroach allows `SELECT 1 = ANY(1, 2)` while Postgres does not. This is a downside of typing subqueries as `TTuple`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 7 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L21yUEp6TQo2IlgMGry:b-cgvkhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Subqueries are typed as `TTuple`. The type checking code allows `{ANY,SOME,ALL} (<tuple>)` in order to support `SELECT 1 = ANY(SELECT 1, 2)`. 

Note that cockroach allows `SELECT 1 = ANY(1, 2)` while Postgres does not. This is a downside of typing subqueries as `TTuple`.
</blockquote></details>

Right. My question is when/why would this be TArray and not TTuple?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 7 of 16 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L22-tYQ853GYpdmj02d:bkv8601) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Right. My question is when/why would this be TArray and not TTuple?
</blockquote></details>

For `SELECT 1 = ANY(ARRAY[1,2])`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 4 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L221PqA2dmcFJy7Om2x:b-hyrejd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

For `SELECT 1 = ANY(ARRAY[1,2])`.
</blockquote></details>

Oh, I didn't know that's valid. Cool, thanks. Added comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Fixed the message. I rebased and fixed some conflicts in the first commmit.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-:-L224lqJGKOrKa2nE8AC:bm5qi46)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L225ZnzTpPKHobyZ0ke:b-5b33hj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh, I didn't know that's valid. Cool, thanks. Added comments.
</blockquote></details>

Cool thanks.  I'm probably missing something, but why is `rightRet` set to `t[0]`?  Shouldn't it be a type?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L226YyRFI5HjvJsX8E3:b-myg81q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Cool thanks.  I'm probably missing something, but why is `rightRet` set to `t[0]`?  Shouldn't it be a type?
</blockquote></details>

The original expression was typed as `<left> <subop> <op> <tuple{T}>`. `foldComparisonExpr` is transforming returning `fOp` (aka `<subop>`) as something like `EQ`, `GT`, `LT`. Below, we're verifying that we have a comparison operator for `<left> <subop> <T>`.

Note that something similar is being done for array, where `<left> <subop> <op> <array[T]>` is being verified below as `<left> <subop> <T>`.

This does point out, though, that we should be verifying all of the types in the tuple are the same. There is a method to do this somewhere. Perhaps `TypeCheckSameTypedExprs`. Maybe. I'm not terribly familiar with this code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L22782fIyvIa7uajTFZ:b35rs4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
What feels off is that a subquery could effectively become the tuple `(1)`, or the tuple `(1, 2)` etc (each of these having a different type). There is a TODO about this here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/subquery.go#L450

> This does point out, though, that we should be verifying all of the types in the tuple are the same

Wouldn't that happen during type-checking? This function is only called on type-checked expressions AFAICT.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L229RwQ2d76I24-nyA1:b-y9vxc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What feels off is that a subquery could effectively become the tuple `(1)`, or the tuple `(1, 2)` etc (each of these having a different type). There is a TODO about this here: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/subquery.go#L450

> This does point out, though, that we should be verifying all of the types in the tuple are the same

Wouldn't that happen during type-checking? This function is only called on type-checked expressions AFAICT.
</blockquote></details>

Thanks for the info, @petermattis!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L229msdKmplHg4hjLBh:b-8nhal4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
> Wouldn't that happen during type-checking? This function is only called on type-checked expressions AFAICT.

Yes, it should, though couldn't this function be used to build an invalid typed expr.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L22Ccri50-A_pY3Gv6_:b2vzuii) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> Wouldn't that happen during type-checking? This function is only called on type-checked expressions AFAICT.

Yes, it should, though couldn't this function be used to build an invalid typed expr.
</blockquote></details>

In cases like `x = ANY(NULL,1, 2)`, the first one has type NULL. And it's not as easy as picking the first non-NULL, because of tuples in tuples `ANY((1, NULL), (NULL, 2))`. As for using `TypeCheckSameTypedExprs`, I'm not sure performing the type-checking for all expressions is what we want. In any case, there is no general way to get a list of expressions (in fact, there is no such thing for a `subquery` node). So we would need the recursive logic in `TypeCheckSameTypedExprs` to be reimplemented to work on types instead of exprs.

I can file an issue after merging to look into this more. I don't think it should hold up this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/expr.go, line 439 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217#-L21sWeEFbrZPskYikjU:-L22Ff_pG-Ec3ms5epMG:b-jkltee) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ab198b01562e979a6dfc5e73ee7c2a2364bc96e/pkg/sql/sem/tree/expr.go#L439)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In cases like `x = ANY(NULL,1, 2)`, the first one has type NULL. And it's not as easy as picking the first non-NULL, because of tuples in tuples `ANY((1, NULL), (NULL, 2))`. As for using `TypeCheckSameTypedExprs`, I'm not sure performing the type-checking for all expressions is what we want. In any case, there is no general way to get a list of expressions (in fact, there is no such thing for a `subquery` node). So we would need the recursive logic in `TypeCheckSameTypedExprs` to be reimplemented to work on types instead of exprs.

I can file an issue after merging to look into this more. I don't think it should hold up this PR.
</blockquote></details>

:thumbsup: I don't think this should hold up this PR either. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21217)*
<!-- Sent from Reviewable.io -->
